### ISSUE TITLE
@cached is "not a decorator" when used in a v2 addon being imported by an Ember 3.28 app with ember-auto-import  2.4.2

### ISSUE THREAD

**Issue by DingoEatingFuzz** [ef4/ember-auto-import#536]
Apologies for the mouthful of a title, it's a very specific thing 😅 

[Reproduction repo](https://github.com/DingoEatingFuzz/cached-with-ember-3) Just run `yarn start:test-app` to see the bug happen at runtime on the application route.

## Setup

This will happen with any Ember 3.28 app importing a V2 addon that uses the `cached-decorator-polyfill`. In the reproduction above, this is shown in a monorepo for convenience, but it'll also happen with published npm packages and independent app repos, as you'd imagine.

1. Create a v2 addon
2. `yarn add ember-cached-decorator-polyfill`
3. Use `import { cached } from '@glimmer/tracking'` in any exported module

## Steps to Reproduce

1. `yarn add your-v2-addon` in any ember app
2. Import the module that uses `@cached` (directly or indirectly through a component/helper/service)
3. Receive `Uncaught TypeError: decorator is not a function` from the bundled module that uses `@cached`

This does not happen with Ember 4.6.x, presumably because the cached polyfill is no longer needed. This also does not happen when building the Ember 3.28 app with embroider.

Since v2 addons are meant to be backwards compatible, I still think this is a bug somewhere in the ember-auto-import or embroider compat pipeline.

**Comment by ef4** [comment]
May be caused by https://github.com/ef4/ember-auto-import/issues/504

**Comment by ef4** [comment]
After looking closer, I think this is because the cached decorator polyfill needs changes to make it work correctly in v2 addons.

The polyfill hasn't been released since before the ember version that shipped the real implementation of `cached`, so it's actually always enabled. It should disable itself on ember >= 4.1.

But in order to do that inside a v2 addon, v2 addons should add the polyfill's babel plugin to their own (prepublish) build and the babel plugin should be updated to emit macros.

Right now it unconditionally rewrites:

```diff
-import { cached } from '@glimmer/tracking';
+import { cached } from 'ember-cached-decorator-polyfill';
```

We want it instead to do:

```diff
-import { cached } from '@glimmer/tracking';
+import { macroCondition, dependencySatisfies, importSync } from '@embroider/macros';
+let cached;
+if (macroCondition(dependencySatisfies('ember-source', '>= 4.1.0-alpha.0'))) {
+  cached = importSync('@glimmer/tracking).cached;
+} else {
+  cached = importSync('ember-cached-decorator-polyfill').cached;
+}
```

**Comment by NullVoxPopuli** [comment]
It looks like this has been applied in: https://github.com/ember-polyfills/ember-cached-decorator-polyfill/pull/184/files

But hasn't been released?

**Comment by ef4** [comment]
Should be fixed in ember-cached-decorator 1.0.1.

**Comment by NullVoxPopuli** [comment]
I confirm this is true
